src/compiler/scala/tools/nsc/transform/patmat/MatchTranslation.scala

```diff
@@ -453,14 +453,16 @@ trait MatchTranslation {
       protected def lengthGuard(binder: Symbol): Option[Tree] =
         // no need to check unless it's an unapplySeq and the minimal length is non-trivially satisfied
         checkedLength map { expectedLength =>
-          def lengthCompareSym = binder.info member nme.lengthCompare
-
           // `binder.lengthCompare(expectedLength)`
           // ...if binder has a lengthCompare method, otherwise
           // `scala.math.signum(binder.length - expectedLength)`
-          def checkExpectedLength = lengthCompareSym match {
-            case NoSymbol => compareInts(Select(seqTree(binder, forceImmutable = false), nme.length), LIT(expectedLength))
-            case lencmp   => (seqTree(binder, forceImmutable = false) DOT lencmp)(LIT(expectedLength))
+          def checkExpectedLength = {
+            val tree = seqTree(binder, forceImmutable = false)
+            val lengthCompareSym = tree.symbol.info.member(nme.lengthCompare)
```

Comment:
Interesting! I think we should try to have `seqTree` produce a typed tree, so that we can do `tree.tpe.member(nme.lengthCompare)`. It seems to already do that in most cases, but not quite. It also doesn't produce a tree with a symbol (`Select(ref, nme.toSeq)`)

Reply:
Yeah, that sounds a bit safer. Will give it a shot. The `Select(ref, nme.toSeq)` case actually doesn't flow in here because we're of `forceImmutable = false`.

Reply:
I managed to get things working by typing the result of `seqTree` within this method.